### ISSUE TITLE
Personalized results

### ISSUE THREAD

**Issue by sradevski** [michaelrambeau/bestofjs-webui#41]
Hey Michael!

One killer feature would be to be able to add projects I am interested in or use, and I can see the state of each project on a personal page (like my projects or something). Many times I keep a list of libraries/projects that I use and silently some of them get outdated, or stop getting maintained, and I want to be aware when something like this happens, or when a project loses popularity. 

For a start I think the same view as for Popular/Trending can be used, but personalized (and optionally private) for me. This can be a simple list of URLs that I specify, and I get back the stats for each of the projects.

What do you think? I am a bit busy for the next two weeks, but after that I can help out with the implementation.

**Comment by michaelrambeau** [comment]
Hello Steve, what an interesting idea, thank you.
The application would be more valuable if we could deliver a customized experience indeed.

We could add a button "Add to my projects" on every project page once a user is logged in.
We could also scan user's projects on Github and extract information from the `package.json` file.

E.g.

```
Dependencies found in your project:
* React (found in 10 projects)
* Express (found in 3 projets)
* ...
```

We could also add a part "The projects I contribute for", finding if the logged-in user has contributed to some of the projects mentioned in the app.

I'm going to make some tests to see if we could extract interesting data from user's `package.json` files.

**Comment by sradevski** [comment]
@michaelrambeau I was also thinking of using the package.json as a source, but didn't mention it in order to keep things simple for a start. Scanning the github repository is definitely a way to go, but for a start just creating a customized page with "Add to my projects" button would suffice I think.

**Comment by michaelrambeau** [comment]
@sradevski  I definitively agree with you, it's better to focus on a simple feature first.

So let's start with a simple page called "My projects", that will be filled with projects that users bookmark using the  "Add to my projects" button.
Then, the next step could be fetching the projects the users have starred on GitHub.
(it's simpler than scanning user's projects package.json)

**Comment by michaelrambeau** [comment]
## Thinking about the implementation

To store user's bookmarked projects, we could use user's "metadata" from Auth0 API.

https://auth0.com/docs/api/management/v2#!/Users/get_users_by_id

There are 2 kinds of metadata stored in user's profile.
According to the Auth0 dashboard:

* `user_metadata`:  Data that the user has read/write access to (e.g. color_preference, blog_url, etc.)
* `app_metadata`:  Data that the user has read-only access to (e.g. roles, permissions, vip, etc)

So we could use `/api/v2/users/{id}` end point to read and update user's bookmarked projects.

**Comment by michaelrambeau** [comment]
## First iteration

I implemented the very first version, it can be accessed from http://bestofjs.netlify.com/
(it's a kind of staging environment to test new features)

### Points

* Users can bookmark projects by using the new `ADD` button at the at the bottom of the project "cards"
* Bookmarked projects can accessed from the new "MY PROJECTS" page, added to the sidebar.
* Bookmarked projects are stored inside user's Auth0 profile, using `usermetadata` field to store an array of project "slugs". No backend to implement :). 

```json
{
  "projects": [
    "react",
    "you-dont-know-js",
    "semanticui",
    "create-react-app",
    "keystone"
  ]
}
```

### Screenshots

![image](https://user-images.githubusercontent.com/5546996/27990738-b9a29052-649b-11e7-8469-ad3d94a6ec08.png)

![image](https://user-images.githubusercontent.com/5546996/27990751-36c8dd8e-649c-11e7-8dac-317203470206.png)

### Feedback?

@steve Is it what you expected? 
I'd like to know your feedback about the wording: the page title, the button labels...

* "My projects", "My bookmarked projects"?
* "Add to my projects",  "Bookmark"?
* "Remove from my projects", "Remove bookmark"?

**Comment by sradevski** [comment]
@michaelrambeau That was really fast, thank you for the effort! I think it is a great first version. What I just noticed when I started using it is, on different projects I use different libraries, and I might be interested to see the library dependencies per project. I think we can create a sub-tree under my projects and let the user assign tags, something similar to the existing structure (only customized per user). 

I think this is outside the scope of this issue, so after checking the "My projects" feature in more detail, we can close this issue and create new ones for reading from package.json and allowing for the user to create tags per project.

Once again, amazing work!

**Comment by michaelrambeau** [comment]
Hello Steve @sradevski ,
Thank you for your comments.
About user's custom tags, I see what you mean, I'd say it's a bit more complicated to implement, compared to what I did, because we would have to store and maintain lists of tags by user.

Also, maybe we could let the user sort "His/her projects" :

* by popularity (total number of stars): this is the current sort
* by trend
* by date (when the bookmark has been added)

...which means I think we should store some "meta" information about the bookmarked project, in the user's profile, starting with `"bookmarked_at`:

```json
{
  "projects": [
    {
      "slug": "react",
      "bookmarked_at": "2017-07-11T22:02:14.179Z"
    }
  ]
}
```

This way, we would be able to store user's custom tags, if needed.
